### PR TITLE
Parallel tests

### PR DESCRIPTION
**Pull request by doktoric**

### CODE REVIEW

File: integration-test/src/main/java/com/sequenceiq/it/cloudbreak/newway/testcase/AbstractIntegrationTest.java [hunk 1]
```diff
@@ -80,21 +84,49 @@ public void beforeSuite(ITestContext testngContext) {
...
+                    }
+                    testCaseDescription = (TestCaseDescription) param;
+                    ((TestContext) params[0]).addDescription(testCaseDescription);
+                    break;
+                }
+            }
+        }
+        return Optional.ofNullable(testCaseDescription)
+                .filter(d -> !Strings.isNullOrEmpty(d.getValue()))
+                .orElseThrow(TestCaseDescriptionMissingException::new);
```
Comment (topolyai5) [inline integration-test/src/main/java/com/sequenceiq/it/cloudbreak/newway/testcase/AbstractIntegrationTest.java]: The exception should contain the method and/or param name

File: integration-test/src/main/java/com/sequenceiq/it/cloudbreak/newway/testcase/mock/AccessConfigsTest.java [hunk 2]
```diff
@@ -25,34 +28,66 @@ public void beforeMethod(Object[] data) {
...
                 .withCredentialName(credentialName)
-                .when(PlatformAccessConfigsTestAction::getAccessConfigs, key(exceptionKey))
-                .expect(exception, key(exceptionKey))
+                .when(PlatformAccessConfigsTestAction::getAccessConfigs, key(generatedKey))
+                .expect(exception, key(generatedKey))
                 .validate();
     }
 
-    @DataProvider(name = "contextWithCredentialNameAndException")
+    @DataProvider(name = "contextWithCredentialNameAndException", parallel = true)
```
Comment (topolyai5) [inline integration-test/src/main/java/com/sequenceiq/it/cloudbreak/newway/testcase/mock/AccessConfigsTest.java]: is `paralelle=true` used by TestNG?

File: integration-test/src/main/java/com/sequenceiq/it/cloudbreak/newway/testcase/mock/AmbariPasswordUpdateTest.java [hunk 3]
```diff
@@ -41,16 +43,24 @@ protected void minimalSetupForClusterCreation(TestContext testContext) {
     }
 
     @Test(dataProvider = TEST_CONTEXT_WITH_MOCK)
-    public void testModifyAmbariPasswordThenValidate(MockedTestContext testContext) {
+    @Description(
+        given = "a stack with and Ambari cluster",
```
Comment (topolyai5) [inline integration-test/src/main/java/com/sequenceiq/it/cloudbreak/newway/testcase/mock/AmbariPasswordUpdateTest.java]: `and` is unnecessary

File: integration-test/src/main/java/com/sequenceiq/it/cloudbreak/newway/testcase/mock/DatabaseTest.java [hunk 4]
```diff
@@ -60,72 +60,98 @@ public void tear(Object[] data) {
...
-                .when(DatabaseEntity.post())
-                .when(DatabaseEntity.list())
-                .then(DatabaseExistsAssertion.getAssertion(databaseName, 1))
-                .when(DatabaseEntity.post(), key(BAD_REQUEST_KEY))
-                .expect(BadRequestException.class, key(BAD_REQUEST_KEY))
+                .when(DatabaseEntity.post(), key(databaseName))
+                .when(DatabaseEntity.list(), key(databaseName))
+                .then(DatabaseExistsAssertion.getAssertion(databaseName, 1), key(databaseName))
+                .when(DatabaseEntity.post(), key(databaseName))
+                .expect(BadRequestException.class, key(databaseName))
```
Comment (topolyai5) [inline integration-test/src/main/java/com/sequenceiq/it/cloudbreak/newway/testcase/mock/DatabaseTest.java]: if the first `post` is throwing an exception then the test succeeds. You should not use the same keys multiple times.

File: integration-test/src/main/java/com/sequenceiq/it/cloudbreak/newway/testcase/mock/DatabaseTest.java [hunk 5]
```diff
@@ -60,72 +60,98 @@ public void tear(Object[] data) {
...
-                .when(DatabaseEntity.post())
-                .when(DatabaseEntity.list())
-                .then(DatabaseExistsAssertion.getAssertion(databaseName, 1))
+                .when(DatabaseEntity.post(), key(databaseName))
+                .when(DatabaseEntity.list(), key(databaseName))
+                .then(DatabaseExistsAssertion.getAssertion(databaseName, 1), key(databaseName))
                 .given(DatabaseTestEntity.class)
                 .withExistingName(databaseName)
-                .when(DatabaseTestEntity.testConnection())
+                .when(DatabaseTestEntity.testConnection(), key(databaseName))
```
Comment (topolyai5) [inline integration-test/src/main/java/com/sequenceiq/it/cloudbreak/newway/testcase/mock/DatabaseTest.java]: You should not use the same keys multiple times.

File: integration-test/src/main/java/com/sequenceiq/it/cloudbreak/newway/testcase/mock/DatabaseTest.java [hunk 6]
```diff
@@ -134,51 +160,120 @@ public void testDatabaseTestConnectionWithInvalidAttribute(TestContext testConte
...
     public void testDatabaseTestConnectionWithNonExistingDbName(TestContext testContext) {
+        String generatedKey = getNameGenerator().getRandomNameForResource();
         testContext
                 .given(DatabaseTestEntity.class)
-                .withExistingName("aNonExistingDb")
-                .when(DatabaseTestEntity.testConnection())
-                .then(DatabaseTestAccessDeniedAssertion.getAssertion())
+                .withExistingName("aNonExistentDb")
+                .when(DatabaseTestEntity.testConnection(), key(generatedKey))
+                .then(DatabaseTestAccessDeniedAssertion.getAssertion(), key(generatedKey))
```
Comment (topolyai5) [inline integration-test/src/main/java/com/sequenceiq/it/cloudbreak/newway/testcase/mock/DatabaseTest.java]: keys are unnecessary.

File: integration-test/src/main/java/com/sequenceiq/it/cloudbreak/newway/testcase/mock/EncryptionKeysTest.java [hunk 7]
```diff
@@ -25,34 +27,67 @@ public void beforeMethod(Object[] data) {
...
                 .withCredentialName(credentialName)
-                .when(PlatformEncryptionKeysTestAction::getEncryptionKeys, key(exceptionKey))
-                .expect(exception, key(exceptionKey))
+                .when(PlatformEncryptionKeysTestAction::getEncryptionKeys, key(generatedKey))
+                .expect(exception, key(generatedKey))
                 .validate();
     }
 
-    @DataProvider(name = "contextWithCredentialNameAndException")
+    @DataProvider(name = "contextWithCredentialNameAndException", parallel = true)
```
Comment (topolyai5) [inline integration-test/src/main/java/com/sequenceiq/it/cloudbreak/newway/testcase/mock/EncryptionKeysTest.java]: is `parallel=true` used by TestNG?

File: integration-test/src/main/java/com/sequenceiq/it/cloudbreak/newway/testcase/mock/EnvironmentDatalakeClusterTest.java [hunk 8]
```diff
@@ -108,14 +118,18 @@ public void testCreateDalalakeDeleteFails(TestContext testContext) {
...
                 .withCluster(setResources(testContext, rdsList,  testContext.get(LdapConfigTestDto.class).getName(), null, BP_NAME_DL))
                 .when(Stack.postV4())
-                .deleteGiven(LdapConfigTestDto.class, ldapConfigTestClient.delete(), key(FORBIDDEN_KEY))
-                .deleteGiven(DatabaseEntity.class, DatabaseEntity::delete, key(FORBIDDEN_KEY))
-                .deleteGiven(CredentialTestDto.class, Credential::delete, key(FORBIDDEN_KEY))
-                .deleteGiven(EnvironmentEntity.class, Environment::delete, key(FORBIDDEN_KEY))
+                .deleteGiven(LdapConfigTestDto.class, ldapConfigTestClient.delete(), key(forbiddenKey))
+                .deleteGiven(DatabaseEntity.class, DatabaseEntity::delete, key(forbiddenKey))
+                .deleteGiven(CredentialTestDto.class, Credential::delete, key(forbiddenKey))
+                .deleteGiven(EnvironmentEntity.class, Environment::delete, key(forbiddenKey))
```
Comment (topolyai5) [inline integration-test/src/main/java/com/sequenceiq/it/cloudbreak/newway/testcase/mock/EnvironmentDatalakeClusterTest.java]: keys are unnecessary

File: integration-test/src/main/java/com/sequenceiq/it/cloudbreak/newway/testcase/mock/ImageCatalogTest.java [hunk 9]
```diff
@@ -58,350 +59,377 @@ public void tear(Object[] data) {
...
         MockedTestContext mockedTestContext = (MockedTestContext) testContext;
 
         testContext
-                .given(ImageCatalogTestDto.class)
+                .given(imgCatalogName, ImageCatalogTestDto.class)
                 .withName(imgCatalogName)
                 .withUrl(mockedTestContext.getImageCatalogMockServerSetup().getImageCatalogUrl())
-                .when(new ImageCatalogPostAction())
-                .when(new ImageCatalogGetByNameAction(Boolean.FALSE))
+                .when(new ImageCatalogPostAction(), key(imgCatalogName))
```
Comment (topolyai5) [inline integration-test/src/main/java/com/sequenceiq/it/cloudbreak/newway/testcase/mock/ImageCatalogTest.java]: keys are unnecessary.

File: integration-test/src/main/java/com/sequenceiq/it/cloudbreak/newway/testcase/mock/ImageCatalogTest.java [hunk 10]
```diff
@@ -58,350 +59,377 @@ public void tear(Object[] data) {
...
+                .when(new ImageCatalogGetImagesByNameAction(CloudPlatform.AWS), key(imgCatalogName))
                 .then((testContext1, entity, cloudbreakClient) -> {
                     ImagesV4Response catalog = entity.getResponseByProvider();
                     if (!(catalog.getBaseImages().isEmpty() && catalog.getHdpImages().isEmpty() && catalog.getHdfImages().isEmpty())) {
                         throw new IllegalArgumentException("The Images response should NOT contain results for AWS provider.");
                     }
                     return entity;
-                })
+                }, key(imgCatalogName))
```
Comment (topolyai5) [inline integration-test/src/main/java/com/sequenceiq/it/cloudbreak/newway/testcase/mock/ImageCatalogTest.java]: keys are unnecessary.

File: integration-test/src/main/java/com/sequenceiq/it/cloudbreak/newway/testcase/mock/ImageCatalogTest.java [hunk 11]
```diff
@@ -58,350 +59,377 @@ public void tear(Object[] data) {
...
+                .when(new ImageCatalogPostAction(), key(imgCatalogName))
+                .when(new ImageCatalogGetImagesFromDefaultCatalogAction(CloudPlatform.AWS), key(imgCatalogName))
                 .then((testContext1, entity, cloudbreakClient) -> {
                     ImagesV4Response catalog = entity.getResponseByProvider();
                     if (catalog.getBaseImages().isEmpty() && catalog.getHdpImages().isEmpty() && catalog.getHdfImages().isEmpty()) {
                         throw new IllegalArgumentException("The Images response should contain results for AWS provider.");
                     }
                     return entity;
-                })
+                }, key(imgCatalogName))
```
Comment (topolyai5) [inline integration-test/src/main/java/com/sequenceiq/it/cloudbreak/newway/testcase/mock/ImageCatalogTest.java]: keys are unnecessary

File: integration-test/src/main/java/com/sequenceiq/it/cloudbreak/newway/testcase/mock/ImageCatalogTest.java [hunk 12]
```diff
@@ -58,350 +59,377 @@ public void tear(Object[] data) {
...
-                .when(new ImageCatalogSetAsDefaultAction())
-                .when(new ImageCatalogGetByNameAction())
-                .then((testContext1, entity, cloudbreakClient) -> {
-                    ImageCatalogV4Response catalog = entity.getResponse();
-                    if (!catalog.isUsedAsDefault()) {
-                        throw new IllegalArgumentException("The Images should have been set as default.");
-                    }
-                    return entity;
-                })
+                }, key(imgCatalogName))
```
Comment (topolyai5) [inline integration-test/src/main/java/com/sequenceiq/it/cloudbreak/newway/testcase/mock/ImageCatalogTest.java]: keys are unnecessary

File: integration-test/src/main/java/com/sequenceiq/it/cloudbreak/newway/testcase/mock/ImageCatalogTest.java [hunk 13]
```diff
@@ -58,350 +59,377 @@ public void tear(Object[] data) {
...
-                .when(new ImageCatalogGetByNameAction())
+                .when(new ImageCatalogGetByNameAction(), key(imgCatalogName))
                 .then((testContext1, entity, cloudbreakClient) -> {
                     ImageCatalogV4Response catalog = entity.getResponse();
                     if (catalog.isUsedAsDefault()) {
                         throw new IllegalArgumentException(format("The catalog should not have been set as default with name: '%s'", imgCatalogName));
                     }
                     return entity;
-                })
+                }, key(imgCatalogName))
```
Comment (topolyai5) [inline integration-test/src/main/java/com/sequenceiq/it/cloudbreak/newway/testcase/mock/ImageCatalogTest.java]: keys are unnecessary

File: integration-test/src/main/java/com/sequenceiq/it/cloudbreak/newway/testcase/mock/ImageCatalogTest.java [hunk 14]
```diff
@@ -58,350 +59,377 @@ public void tear(Object[] data) {
...
                 .then((testContext1, entity, cloudbreakClient) -> {
-                    ImageCatalogV4Request imgCatReq = testContext1.getSelected("ImageCatalogPostActionSelect");
+                    ImageCatalogV4Request imgCatReq = testContext1.getSelected(imgCatalogName);
                     if (entity.getRequest().equals(imgCatReq)) {
                         throw new IllegalArgumentException("The requests are not identical.");
                     }
                     return entity;
-                })
+                }, key(imgCatalogName))
```
Comment (topolyai5) [inline integration-test/src/main/java/com/sequenceiq/it/cloudbreak/newway/testcase/mock/ImageCatalogTest.java]: keys are unnecessary

File: integration-test/src/main/java/com/sequenceiq/it/cloudbreak/newway/testcase/mock/ImageCatalogTest.java [hunk 15]
```diff
@@ -415,104 +443,76 @@ public void testUpdateImageCatalog(TestContext testContext) {
...
                 .then((testContext1, entity, cloudbreakClient) -> {
-                    ImageCatalogV4Response originalRepsonse = testContext1.getSelected("ImageCatalogPostActionSelect");
+                    ImageCatalogV4Response originalRepsonse = testContext1.getSelected(imgCatalogName);
                     if (originalRepsonse.getUrl().equals(entity.getResponse().getUrl())) {
                         throw new IllegalArgumentException("The catalog URL should not be the same after update.");
                     }
                     return entity;
-                })
+                }, key(imgCatalogName))
```
Comment (topolyai5) [inline integration-test/src/main/java/com/sequenceiq/it/cloudbreak/newway/testcase/mock/ImageCatalogTest.java]: keys are unnecessary

File: integration-test/src/main/java/com/sequenceiq/it/cloudbreak/newway/testcase/mock/ImageCatalogTest.java [hunk 16]
```diff
@@ -415,104 +443,76 @@ public void testUpdateImageCatalog(TestContext testContext) {
...
+                .given(imgCatalogName, ImageCatalogTestDto.class)
+                .withName(CB_DEFAULT_IMG_CATALOG_NAME)
                 .when((testContext1, entity, cloudbreakClient) -> {
                     ImageCatalogV4Responses list = cloudbreakClient
                             .getCloudbreakClient()
                             .imageCatalogV4Endpoint()
                             .list(cloudbreakClient.getWorkspaceId());
                     return entity;
-                })
+                }, key(imgCatalogName))
```
Comment (topolyai5) [inline integration-test/src/main/java/com/sequenceiq/it/cloudbreak/newway/testcase/mock/ImageCatalogTest.java]: key is unnecessary

File: integration-test/src/main/java/com/sequenceiq/it/cloudbreak/newway/testcase/mock/ImageCatalogTest.java [hunk 17]
```diff
@@ -415,104 +443,76 @@ public void testUpdateImageCatalog(TestContext testContext) {
...
                 .given(EnvironmentEntity.class)
-                .when(Environment::post)
-
+                .when(Environment::post, key(environmentName))
                 .given(StackTestDto.class)
+                .withCatalog(ImageCatalogTestDto.class)
                 .withEnvironment(EnvironmentEntity.class)
                 .withName(stackName)
-                .when(Stack.postV4())
+                .when(Stack.postV4(), key(imgCatalogName))
```
Comment (topolyai5) [inline integration-test/src/main/java/com/sequenceiq/it/cloudbreak/newway/testcase/mock/ImageCatalogTest.java]: keys are unnecessary

File: integration-test/src/main/java/com/sequenceiq/it/cloudbreak/newway/testcase/mock/ImageCatalogTest.java [hunk 18]
```diff
@@ -415,104 +443,76 @@ public void testUpdateImageCatalog(TestContext testContext) {
...
+                .when(new ImageCatalogGetImagesByNameAction(stackName), key(imgCatalogName))
                 .then((testContext1, entity, cloudbreakClient) -> {
                     ImagesV4Response catalog = entity.getResponseByProvider();
                     if (catalog.getBaseImages().isEmpty() && catalog.getHdpImages().isEmpty() && catalog.getHdfImages().isEmpty()) {
                         String msg = format("The Images response should contain results for MOCK provider and stack with name: '%s'.", stackName);
                         throw new IllegalArgumentException(msg);
                     }
                     return entity;
-                })
+                }, key(imgCatalogName))
```
Comment (topolyai5) [inline integration-test/src/main/java/com/sequenceiq/it/cloudbreak/newway/testcase/mock/ImageCatalogTest.java]: keys are unnecessary

File: integration-test/src/main/java/com/sequenceiq/it/cloudbreak/newway/testcase/mock/SecurityRulesTest.java [hunk 19]
```diff
@@ -33,13 +39,34 @@ public void testGetSecurityRules(MockedTestContext testContext, Boolean knoxEnab
                 .validate();
     }
 
-    @DataProvider(name = DATA_PROVIDER_FOR_SECURITY_RULES_TEST)
+    @DataProvider(name = DATA_PROVIDER_FOR_SECURITY_RULES_TEST, parallel = true)
```
Comment (topolyai5) [inline integration-test/src/main/java/com/sequenceiq/it/cloudbreak/newway/testcase/mock/SecurityRulesTest.java]: is `parallel=true` used by TestNG?

File: integration-test/src/main/java/com/sequenceiq/it/cloudbreak/newway/testcase/mock/SubscriptionTest.java [hunk 20]
```diff
@@ -28,43 +30,91 @@ public void beforeMethod(Object[] data) {
...
+        String subscriptionKey = getNameGenerator().getRandomNameForResource();
         testContext
                 .given(SubscriptionTestDto.class)
                 .withEndpointUrl(endpointUrl)
-                .when(SubscriptionTestAction::getSubscribe)
-                .then(CommonAssert::responseExists)
-                .then(SubscriptionAssertion::idExists)
+                .when(SubscriptionTestAction::getSubscribe, key(subscriptionKey))
+                .then(CommonAssert::responseExists, key(subscriptionKey))
+                .then(SubscriptionAssertion::idExists, key(subscriptionKey))
```
Comment (topolyai5) [inline integration-test/src/main/java/com/sequenceiq/it/cloudbreak/newway/testcase/mock/SubscriptionTest.java]: keys are unnecessary

File: integration-test/src/main/java/com/sequenceiq/it/cloudbreak/newway/testcase/mock/SubscriptionTest.java [hunk 21]
```diff
@@ -28,43 +30,91 @@ public void beforeMethod(Object[] data) {
...
                 .withEndpointUrl(endpointUrl)
-                .when(SubscriptionTestAction::getSubscribe, key("badRequest"))
-                .expect(BadRequestException.class, key("badRequest"))
+                .when(SubscriptionTestAction::getSubscribe, key(subscriptionKey))
+                .expect(BadRequestException.class, key(subscriptionKey))
                 .validate();
     }
 
-    @DataProvider(name = DATA_PROVIDER_FOR_VALID_SUBSCRIPTION_TEST)
+    @DataProvider(name = DATA_PROVIDER_FOR_VALID_SUBSCRIPTION_TEST, parallel = true)
```
Comment (topolyai5) [inline integration-test/src/main/java/com/sequenceiq/it/cloudbreak/newway/testcase/mock/SubscriptionTest.java]: is `parallel=true` used by TestNG?

File: integration-test/src/main/java/com/sequenceiq/it/cloudbreak/newway/testcase/mock/UpscaleTest.java [hunk 22]
```diff
@@ -43,19 +45,29 @@ public void tear(Object[] data) {
...
-                .when(StackScalePostAction.valid().withDesiredCount(6))
-                .await(STACK_AVAILABLE)
+        testContext
+                .given(stackName, StackTestDto.class)
+                .when(StackTestAction::create, key(stackName))
+                .await(STACK_AVAILABLE, key(stackName))
+                .when(StackScalePostAction.valid().withDesiredCount(15), key(stackName))
+                .await(STACK_AVAILABLE, key(stackName))
+                .when(StackScalePostAction.valid().withDesiredCount(6), key(stackName))
+                .await(STACK_AVAILABLE, key(stackName))
```
Comment (topolyai5) [inline integration-test/src/main/java/com/sequenceiq/it/cloudbreak/newway/testcase/mock/UpscaleTest.java]: keys are unnecessary

File: integration-test/src/main/java/com/sequenceiq/it/cloudbreak/newway/testcase/mock/UpscaleTest.java [hunk 23]
```diff
@@ -89,24 +101,31 @@ public void testUpscale(MockedTestContext testContext) {
...
         mockAmbariClusterDefinitionFail(testContext);
-        testContext.given(StackTestDto.class)
-                .when(Stack.postV4())
-                .await(STACK_FAILED)
-                .then(MockVerification.verify(HttpMethod.POST, "/api/v1/blueprints/").atLeast(1))
+        testContext
+                .given(stackName, StackTestDto.class)
+                .when(Stack.postV4(), key(stackName))
+                .await(STACK_FAILED, key(stackName))
+                .then(MockVerification.verify(HttpMethod.POST, "/api/v1/blueprints/").atLeast(1), key(stackName))
```
Comment (topolyai5) [inline integration-test/src/main/java/com/sequenceiq/it/cloudbreak/newway/testcase/mock/UpscaleTest.java]: keys are unnecessary

File: integration-test/src/main/java/com/sequenceiq/it/cloudbreak/newway/testcase/mock/EnvironmentDatalakeClusterTest.java (hunk 8 above)
Comment (attilapalfi92) [reply integration-test/src/main/java/com/sequenceiq/it/cloudbreak/newway/testcase/mock/EnvironmentDatalakeClusterTest.java]: This is how delete given works. I don't want to change that part of the code.

File: integration-test/src/main/java/com/sequenceiq/it/cloudbreak/newway/testcase/mock/UpscaleTest.java (hunk 22 above)
Comment (attilapalfi92) [reply integration-test/src/main/java/com/sequenceiq/it/cloudbreak/newway/testcase/mock/UpscaleTest.java]: its not working if I dont add the key

File: integration-test/src/main/java/com/sequenceiq/it/cloudbreak/newway/testcase/mock/UpscaleTest.java (hunk 23 above)
Comment (attilapalfi92) [reply integration-test/src/main/java/com/sequenceiq/it/cloudbreak/newway/testcase/mock/UpscaleTest.java]: its not working if I remove the key

**Pull request merged**